File: FinbourneLRUCacheJava/src/com/finbourne/LRUCache.java
package com.finbourne;

import java.util.*;
import java.util.stream.Collectors;


/**
 * Normally, we'd have two type arguments here for Key and Values. But since the
 * assignment explicitly states that we can have arbitrary types of key/value pairs AND
 * that this should be a singleton, we are just going to make it Object/Object.
 * We COULD, in theory, include a factory method that takes two Class arguments and getOrCreate's
 * an instance of LRUCache<K, V> with that specific type pair, but then it gets messy when it comes to subtypes
 * and overly heavy on the syntax.
 * Since this is just a test assignment, we are going to keep it simple and stick to Object/Object. But
 * we are going to manually craft the double-linked support list to give the exercise a bit of depth and for
 * optimization. Head is the LRU item, and Tail is the next to evict.
 *
 * Additionally, we are going to mark the methods as "synchronized" to obtain thread safety.
 *
 * Finally, we are going to make it implement the Map interface to see what methods are necessary to implement.
 */
public class LRUCache implements Map<Object, Object> {
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        LRUCache lruCache = (LRUCache) o;
        return capacity == lruCache.capacity && Objects.equals(map, lruCache.map);
    }

    @Override
    public int hashCode() {
        return Objects.hash(map, capacity);
    }

    private class Node {
        Object key;
        Object value;

        Node prev;

        Node next;

        EvictionCallback evictionCallback = null;

        Node(Object key, Object value) {
            this.key = key;
            this.value = value;
        }

        // No need to make these depend on prev/next pointers semantically.
        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Node node = (Node) o;
            return Objects.equals(key, node.key) && Objects.equals(value, node.value);
        }

        @Override
        public int hashCode() {
            return Objects.hash(key, value);
        }
    }

    private Map<Object, Node> map;

    //The default value here should be chosen appropriately.
    private int capacity = 99;
    private Node head = null;
    private Node tail = null;

    private synchronized Node removeTail() {
        if (tail != null) {
            Node toRemove = tail;
            if (toRemove.evictionCallback != null) {
                toRemove.evictionCallback.apply(toRemove.key, toRemove.value);
            }
            map.remove(toRemove.key);
            Node prev = toRemove.prev;
            if (prev != null) {
                prev.next = null;
                tail = prev;
            }
            else {
                tail = null;
                head = null;
            }
            return toRemove;
        }
        else {
            return null;
        }
    }

    private synchronized void moveToTail(Node n) {
        if (n != tail){
            if (n.prev != null) {
                n.prev.next = n.next;
            }
            if (n.next != null) {
                n.next.prev = n.prev;
                if (n.next.prev == null) {
                    head = n.next;
                }
            }
            if (tail != null) {
                tail.next = n;
                n.prev = tail;
            }
            tail = n;
        }
    }

    private synchronized void moveToHead(Node n) {
        if (n != head) {
            if (n.prev != null) {
                n.prev.next = n.next;
                if (n.prev.next == null) {
                    tail = n.prev;
                }
            }
            if (n.next != null) {
                n.next.prev = n.prev;
            }
            if (head != null) {
                head.prev = n;
                n.next = head;
            }
            head = n;
        }
    }


    private LRUCache() {
        this.map = new HashMap<>();
    }

    private static LRUCache instance = null;

    public static synchronized LRUCache getInstance() {
        if (LRUCache.instance == null) {
            LRUCache.instance = new LRUCache();
        }
        return LRUCache.instance;
    }


    @Override
    public synchronized int size() {
        return map.size();
    }

    @Override
    public synchronized boolean isEmpty() {
        return map.isEmpty();
    }

    @Override
    public synchronized boolean containsKey(Object key) {
        return map.containsKey(key);
    }

    @Override
    public synchronized boolean containsValue(Object value) {
        return map.values().stream().anyMatch(node -> node.value.equals(value));
    }

    /**
     * Getting an item means updating its position in the LRU list
     */
    @Override
    public synchronized Object get(Object key) {
        if (map.containsKey(key)) {
            Node n = map.get(key);
            moveToHead(n);
            return n.value;
        }
        else {
            return null;
        }
    }

    /**
     * Putting a k/v pair means adding it to the map, as well as fixing the pointers of the LRU list.
     * Can also register the eviction callback.
     */
    @Override
    public synchronized Object put(Object key, Object value) {
        return put(key, value, null);
    }

    public synchronized Object put(Object key, Object value, EvictionCallback ec) {
        Object previousValue = get(key);

        Node n = new Node(key, value);
        n.evictionCallback = ec;
        moveToHead(n);
        if (isEmpty()) {
            // If first item it's both head and tail
            moveToTail(n);
        }
        map.put(key, n);

        while(size() > capacity) {
            removeTail();
        }

        return previousValue;
    }

    @Override
    public synchronized Object remove(Object key) {
        if (map.containsKey(key)) {
            Node n = map.remove(key);
            moveToTail(n);
            removeTail();
            return n.value;
        }
        else {
           return null;
        }
    }

    @Override
    public synchronized void putAll(Map<?, ?> m) {
        for(Entry<?, ?> e: m.entrySet()) {
            put(e.getKey(), e.getValue());
        }
    }

    @Override
    public synchronized void clear() {
        map.clear();
        tail = null;
        head = null;
    }

    @Override
    public synchronized Set<Object> keySet() {
        return map.keySet();
    }

    @Override
    public synchronized Collection<Object> values() {
        return map.values().stream().map(node -> node.value).toList();
    }

    @Override
    public synchronized Set<Entry<Object, Object>> entrySet() {
        return map.entrySet()
                .stream()
                .collect(Collectors.toMap(
                        Entry::getKey,
                        e -> e.getValue().value)).entrySet();
    }

    /**
     * Can change the capacity at runtime and return what is evicted.
     */
    public synchronized Collection<Object> resize(int newCapacity) {
        if (newCapacity < 0 ){
            throw new IllegalArgumentException("Cache size can not be negative ("+newCapacity+")");
        }
        else {
            capacity = newCapacity;
            List<Object> removed = new LinkedList<>();
            while(size() > capacity) {
                removed.add(removeTail().value);
            }
            return removed;
        }
    }
}

File: FinbourneLRUCacheCSharp/LRUCache/LRUCache.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LRUCache;

public sealed class LRUCache
{
    private static readonly Lazy<LRUCache> lazy =
        new Lazy<LRUCache>(() => new LRUCache());

    public static LRUCache Instance { get { return lazy.Value; } }

    /// <summary>
    /// The default capacity value should be chosen appropriately, and can be altered using the resize method
    /// </summary>
    private int _capacity = 99;
    private Dictionary<object, Node> _map = new();
    
    //LRU item
    private Node _head = null;
    
    //Next to be evicted
    private Node _tail = null;

    /// <summary>
    /// Lock for all functions accessing the cache
    /// </summary>
    private object _lock = new();

    public Dictionary<object, object> Resize(int newCapacity) {
        lock (_lock)
        {
            if (newCapacity < 0)
            {
                throw new ArgumentException("Cache size can not be negative (" + newCapacity + ")");
            }
            else
            {
                _capacity = newCapacity;
                Dictionary<object, object> removed = new();
                while (Size() > _capacity)
                {
                    Node n = removeTail();
                    removed.Add(n.Key, n.Value);
                }
                return removed;
            }
        }
    }

    private Node removeTail()
    {
        lock (_lock)
        {
            if (_tail != null)
            {
                Node toRemove = _tail;
                if (toRemove.EvictionCallback != null)
                {
                    toRemove.EvictionCallback.Invoke(toRemove.Key, toRemove.Value);
                }
                _map.Remove(toRemove.Key);
                Node prev = toRemove.Prev;
                if (prev != null)
                {
                    prev.Next = null;
                    _tail = prev;
                }
                else
                {
                    _tail = null;
                    _head = null;
                }
                return toRemove;
            }
            else
            {
                return null;
            }
        }
    }

    private void moveToTail(Node n)
    {
        lock (_lock)
        {
            if (n != _tail)
            {
                if (n.Prev != null)
                {
                    n.Prev.Next = n.Next;
                }
                if (n.Next != null)
                {
                    n.Next.Prev = n.Prev;
                    if (n.Next.Prev == null)
                    {
                        _head = n.Next;
                    }
                }
                if (_tail != null)
                {
                    _tail.Next = n;
                    n.Prev = _tail;
                }
                _tail = n;
            }
        }
    }

    private void moveToHead(Node n)
    {
        lock (_lock)
        {
            if (n != _head)
            {
                if (n.Prev != null)
                {
                    n.Prev.Next = n.Next;
                    if (n.Prev.Next == null)
                    {
                        _tail = n.Prev;
                    }
                }
                if (n.Next != null)
                {
                    n.Next.Prev = n.Prev;
                }
                if (_head != null)
                {
                    _head.Prev = n;
                    n.Next = _head;
                }
                _head = n;
            }
        }
    }

    public int Size()
    {
        lock(_lock)
        {
            return _map.Count;
        }
    }

    public bool Empty()
    {
        lock(_lock)
        {
            return Size() == 0;
        }
    }

    public bool ContainsKey(object key)
    {
        lock (_lock)
        {
            return _map.ContainsKey(key);
        }
    }

    public bool ContainsValue(object value)
    {
        lock(_lock)
        {
            return _map.Values.Contains(value);
        }
    }

    /// <summary>
    /// Getting an item means updating its position in the LRU list
    /// </summary>
    public object Get(object key)
    {
        lock (_lock)
        {
            if (_map.ContainsKey(key))
            {
                Node n = _map[key];
                moveToHead(n);
                return n.Value;
            }
            else
            {
                return null;
            }
        }
    }

    public object Put(object key, object value)
    {
        lock (_lock) 
        {
            return Put(key, value, null);
        }
    }

    public object Put(object key, object value, Action<object, object> evictionCallback)
    {
        lock (_lock)
        {
            object previousValue = Get(key);

            Node n = new Node(key, value);
            n.EvictionCallback = evictionCallback;
            moveToHead(n);
            if (Empty())
            {
                // If first item it's both head and tail
                moveToTail(n);
            }
            _map.Add(key, n);

            while (Size() > _capacity)
            {
                removeTail();
            }

            return previousValue;
        }
    }


    public object Remove(object key)
    {
        lock (_lock)
        {
            if (_map.ContainsKey(key))
            {
                Node n = _map[key];
                _map.Remove(key);
                moveToTail(n);
                removeTail();
                return n.Value;
            }
            else
            {
                return null;
            }
        }
    }

    public void PutAll(Dictionary<object, object> m)
    {
        lock(_lock)
        {
            foreach(KeyValuePair<object, object> pair in m)
            {
                Put(pair.Key, pair.Value); 
            }
        }
    }

    public override bool Equals(object? obj)
    {
        return obj is LRUCache cache &&
               _capacity == cache._capacity &&
               EqualityComparer<Dictionary<object, Node>>.Default.Equals(_map, cache._map);
    }

    public void Clear()
    {
        lock (_lock)
        {
            _map.Clear();
            _head = null;
            _tail = null;
        }
    }

    public override int GetHashCode()
    {
        return HashCode.Combine(_capacity, _map);
    }
}